### PR TITLE
Adding WeightSize() to transposed convolution, celu, c_relu, noisy_linear and parametric relu

### PR DESCRIPTION
**Pull request by NippunSharma**
This is in reference to issue #2590

### CODE REVIEW

**Review by Abilityguy — CHANGES_REQUESTED** [review]
There are some style fixes that I have suggested. 
After these changes, the style checks should pass.

**Comment by NippunSharma** [comment]
@Abilityguy  Thanks ! style checks are passed. Also, why is it showing 1 check as errored ?

**Comment by Abilityguy** [comment]
Hey @NippunSharma , I am not sure why that check has failed. I looks like it was automatically cancelled. 
Mostly, it is not related to this.
Let us wait for the other checks to pass.

**Comment by NippunSharma** [comment]
Okay. If you are free can you please also review #2704

**Comment by NippunSharma** [comment]
Sorry, did not know that github automatically links the PR's we mention !
The mention did not go away even after I deleted the comment.

**Comment by Abilityguy** [comment]
Not an issue. :)

**Review by zoq — COMMENTED** [review]
Can you add some tests to https://github.com/mlpack/mlpack/blob/master/src/mlpack/tests/ann_visitor_test.cpp?

**Comment by NippunSharma** [comment]
Hi @zoq , I have added tests for transposed convolution and noisy linear.

**Comment by zoq** [comment]
Hm, for some reason the Static Code Analysis Job failed, so let's restart the build.

**Comment by zoq** [comment]
@mlpack-jenkins test this please

**Comment by NippunSharma** [comment]
@zoq is there any specific reason that this is happening?

**Comment by NippunSharma** [comment]
@mlpack-jenkins test this

**Comment by NippunSharma** [comment]
@zoq Static Code Analysis is now successful

**Comment by zoq** [comment]
> @zoq is there any specific reason that this is happening?

No not really, but I just glanced over the output, looks like a temporary issue.

File: src/mlpack/tests/ann_visitor_test.cpp
Comment (zoq) [inline src/mlpack/tests/ann_visitor_test.cpp:200]: We should delete the layer at the end.

File: src/mlpack/tests/ann_visitor_test.cpp
Comment (zoq) [inline src/mlpack/tests/ann_visitor_test.cpp:217]: See comment above.

File: src/mlpack/tests/ann_visitor_test.cpp
Comment (zoq) [inline src/mlpack/tests/ann_visitor_test.cpp:218]: This should be tabbed twice (4 spaces), see https://github.com/mlpack/mlpack/wiki/DesignGuidelines#line-length-and-wrapping for more details.

**Comment by NippunSharma** [comment]
@zoq okay I will do the specified changes.

**Comment by NippunSharma** [comment]
@zoq I tried to delete the layers but since it is not a pointer it is showing the following error :

`argument given to 'delete', expected pointer
  221 |   delete noisyLinearLayer;
      |          ^~~~~~~~~~~~~~~~`

**Comment by NippunSharma** [comment]
I have added the spaces

**Comment by NippunSharma** [comment]
Hi , @zoq , @Abilityguy , please take a look at this whenever you have time

File: src/mlpack/methods/ann/layer/noisylinear.hpp
Comment (Abilityguy) [inline src/mlpack/methods/ann/layer/noisylinear.hpp:137]:
```suggestion
  size_t WeightSize() const { return (outSize * inSize + outSize) * 2; }

```
A newline here would be good.
It's just a small style fix to retain the general style format across the files.

File: src/mlpack/methods/ann/layer/transposed_convolution.hpp
Comment (Abilityguy) [inline src/mlpack/methods/ann/layer/transposed_convolution.hpp:281]:
```suggestion
  }
  
```
Style fix.

**Review by Abilityguy — CHANGES_REQUESTED** [review]
Hi @NippunSharma , there are a few minor style fixes I have suggested.
Regarding the logic, let's wait for the approval.

**Comment by NippunSharma** [comment]
@Abilityguy , any reason as to why are tests failing ?

**Comment by NippunSharma** [comment]
What is the reason for these failing tests ?

**Comment by NippunSharma** [comment]
Hi @zoq , while building mlpack from source we see the following warning due to some unused variables in the some test files,
regarding the `weightSize` variable, should I go ahead and remove these unused variables in this PR ?
image

**Comment by NippunSharma** [comment]
I believe these variables were meant for the `CheckCorrectnessOfWeightSize` function but are mistakenly initialized outside these functions (These functions do contain a separate `weightSize` variable):
 
image

**Comment by NippunSharma** [comment]
For example here it is defined unnecessarily: 

image

**Comment by NippunSharma** [comment]
Hi @zoq please take a look at this whenever you have time.

**Comment by zoq** [comment]
> For example here it is defined unnecessarily:
> 
> image

Nice catch, in this case let's just remove the output.

File: src/mlpack/tests/ann_visitor_test.cpp
Comment (zoq) [reply src/mlpack/tests/ann_visitor_test.cpp:200]: This is still something we have to do after the `CheckCorrectnessOfWeightSize` check.

Comment (NippunSharma) [reply src/mlpack/tests/ann_visitor_test.cpp:200]: I am not able to to fully understand what are you saying here ? Also I looked at some previous tests such as `WeightSizeVisitorTestForLinearLayer` and the layer is not deleted there also ?

**Comment by NippunSharma** [comment]
@zoq should I create a new PR for this or just add to this one ?

**Comment by zoq** [comment]
> @zoq should I create a new PR for this or just add to this one ?

Let's use this PR.

**Comment by NippunSharma** [comment]
Hi @zoq I have deleted the unused variables. Also, I am not able to completely understand how can I delete the layer at the end ?

File: src/mlpack/tests/ann_visitor_test.cpp
Comment (zoq) [reply src/mlpack/tests/ann_visitor_test.cpp:189]: I guess we missed it there, to delete the layer, we just have to add `delete transposedConvLayer;` at the end.

Comment (NippunSharma) [reply src/mlpack/tests/ann_visitor_test.cpp:189]: `transposedConvLayer` is not a pointer, so I think `delete transposedConvLayer` might lead to error. Should I create it as a pointer ?

Comment (zoq) [reply src/mlpack/tests/ann_visitor_test.cpp:189]: It is a pointer, `LayerTypes<>` holds a pointer.

**Comment by zoq** [comment]
Can you merge in the lastest master branch, that should fix the macOS and Julia build.

File: src/mlpack/tests/ann_visitor_test.cpp
Comment (NippunSharma) [reply src/mlpack/tests/ann_visitor_test.cpp:189]: Yes, I see it now. I will add the changes. Thanks

**Comment by zoq** [comment]
I guess I was wrong and the boost::variant will deconstruct the object for us.

**Comment by NippunSharma** [comment]
@zoq I have removed delete. Please take a look at it.

**Review by zoq — APPROVED** [review]
No more comments from my side, thanks for working on this.

**Comment by NippunSharma** [comment]
@zoq can you merge this ?

**Review by shrit — APPROVED** [review]

**Pull request merged**
